docs: describe how to manage embedded plugin registry for VS Code editor (#2503)
codenvy/che-docs: modules/administration-guide/partials/proc_adding-or-removing-vscode-extensions-in-the-embedded-open-vsx-registry-instance.adoc

:_content-type: PROCEDURE

[id="adding-or-removing-vscode-extensions-in-the-embedded-open-vsx-registry-instance"]
= Adding or removing pass:[<!-- vale RedHat.Spelling = NO -->] pass:[<!-- vale RedHat.Spelling = NO -->] VS Code pass:[<!-- vale RedHat.Spelling = YES -->] pass:[<!-- vale RedHat.Spelling = YES -->] extensions in the embedded Open VSX registry instance

You can add or remove your choice of pass:[<!-- vale RedHat.Spelling = NO -->] pass:[<!-- vale RedHat.Spelling = NO -->] VS Code pass:[<!-- vale RedHat.Spelling = YES -->] pass:[<!-- vale RedHat.Spelling = YES -->] extensions in the embedded Open VSX registry instance that is cached by {prod-short} to support air-gapped, offline, and proxy-restricted environments.

> This will create a custom build of the Open VSX registry, which can be used in your organization’s workspaces.

Reviewer: ```suggestion
This will create a custom build of the Open VSX registry, which can be used in your organization’s workspaces. 

Note that after a subscription update, your CheCluster will revert to default settings and any changes will need to be reapplied. 
```

> Completing this procedure results in an optional, custom build of the embedded Open VSX registry for use in workspaces of your organization’s Che.

Reviewer: ```suggestion
This will create a custom build of the Open VSX registry, which can be used in your organization’s workspaces.
```

.Procedure

include::example$snip_{project-context}-steps-to-add-or-remove-extensions-in-openvsx-sync-json.adoc[]

** If the extension is published on link:https://open-vsx.org[open-vsx.org], you can add it by including an object with the extension `id` in the format __<published_by>__.__<unique_identifier>__ that you can collect from the extension page on link:https://open-vsx.org[open-vsx.org]:
+
[source,json,subs="+quotes"]
----
    {
        "id": "__<published_by>__.__<unique_identifier>__"
    }
----
+
[TIP]
====
The latest extension version on link:https://open-vsx.org[open-vsx.org] is the default. Alternatively, you can add `"version": "__<extension_version>__"` on a new line to specify a version.
====

> ** If it is an extension from link:https://marketplace.visualstudio.com/VSCode[Visual Studio Marketplace] that is not available on link:https://open-vsx.org[Open VSX] yet, you can ask the extension author to publish it on link:https://open-vsx.org[open-vsx.org] according to the link:https://github.com/eclipse/openvsx/wiki/Publishing-Extensions#how-to-publish-an-extension[instructions] so that you can add it as an extension that is published on link:https://open-vsx.org[open-vsx.org].

Reviewer: ```suggestion
** If an extension is only available from link:https://marketplace.visualstudio.com/VSCode[Visual Studio Marketplace], but not link:https://open-vsx.org[Open VSX], you can ask the extension author to publish it on link:https://open-vsx.org[open-vsx.org] according to the link:https://github.com/eclipse/openvsx/wiki/Publishing-Extensions#how-to-publish-an-extension[instructions].

 Otherwise the link:https://aka.ms/vsmarketplace-ToU[Terms Of Use] for the link:https://marketplace.visualstudio.com/VSCode[Visual Studio Marketplace] allow using extensions only in Microsoft's products and services.
```

Reviewer 2: @nickboldt, can I append a phrase at the end of the sentence: "if the extension's publisher is not Microsoft."?

Reviewer: Doesn't matter who the publisher is. It matters that the TOU for the MS registry says "only with In-Scope Products and Services". 

"Marketplace Offerings are intended for use only with In-Scope Products and Services and you may install and use Marketplace Offerings only with In-Scope Products and Services."

The list of products is in paragraph one: "Microsoft Visual Studio, Visual Studio for Mac,
Visual Studio Code, GitHub Codespaces, Azure DevOps, Azure DevOps Server, and successor products
and services (the “In-Scope Products and Services”) offered by us and GitHub, Inc. (“GitHub”)". 

So... no. Just say "don't use them" and link to the TOU.

Reviewer 2: @nickboldt, what I meant is we can't seriously suggest that our customers contact Microsoft with requests to have Microsoft publish its extensions on the Open VSX website.

``` 
If an extension is only available from link:https://marketplace.visualstudio.com/VSCode[Visual Studio Marketplace], but not link:https://open-vsx.org[Open VSX], you can ask the extension author to publish it on link:https://open-vsx.org[open-vsx.org] according to the link:https://github.com/eclipse/openvsx/wiki/Publishing-Extensions#how-to-publish-an-extension[instructions].
```

Reviewer: > ask the extension author

the author != Microsoft in most cases.

Reviewer 2: @nickboldt, can I add `third-party` to it? 
```
** If an extension is only available from ... Visual Studio Marketplace
```
`>`
```
** If a third-party extension is only available from ... Visual Studio Marketplace
```

Reviewer 2: @nickboldt, and I'm also thinking of one more bullet list item. Or we could put it as a TIP. Something like this:

> If there is no Open VSX equivalent for an extension that you use and that is published by Microsoft on Visual Studio Marketplace, consider informing the Open VSX team by filing an issue at https://github.com/open-vsx/publish-extensions/issues.

WDYT?

They do (re)publish some themselves if I understood it correctly from this example: https://open-vsx.org/extension/ms-vscode/cpptools-themes.

Although I can understand that you might not want such a statement to appear in the downstream docs.

Reviewer 2: @nickboldt, OMG:

https://marketplace.visualstudio.com/items?itemName=ms-python.python > https://github.com/Microsoft/vscode-python

![image](https://user-images.githubusercontent.com/78564304/208753272-94c4aa1c-ae4c-46fa-83a7-da342b232ead.png)

I mean the `forked from` part.

Another reason to add my suggestion according to my previous comment above ^.

Reviewer 2: And Microsoft's original extensions have an MIT license: https://github.com/microsoft/vscode-cpptools/blob/main/LICENSE.md. I see it's already been republished into https://open-vsx.org/extension/ms-vscode/cpptools-themes, because clicking the **Repository** link on the https://open-vsx.org page takes me back to Microsoft's repo, https://github.com/microsoft/vscode-cpptools.

Reviewer: Kinda getting into the weeds here. 

If you want a tip in both up/downstream that says something like you suggested above, I'm good with that:

```
If there is no Open VSX equivalent of an extension that you use consider informing the Open VSX team at https://github.com/open-vsx/publish-extensions/issues to let them know.
```

Of course I don't know if the Open VSX maintainers will just rebuild/republish/fork random extensions just because you ask for it... so this might be a fool's errand.

+
[TIP]
====

> If the extension author is not cooperative, you can add the extension directly from the `.vsix` file:

Reviewer: ```suggestion
If the extension author is not cooperative, you can add the extension directly from the `.vsix` file, as long as the license and terms of use allow it, and the file can be accessed from a https:// URL while building the plugin registry container.
```

> If the extension author is not cooperative, you can add the extension directly from the `.vsix` file:

Reviewer: This section should be removed. This implies "if you want something from the https://marketplace.visualstudio.com/VSCode and the author won't publish it to https://open-vsx.org, go ahead and just link to it."

This is 100% in violation of the TOU @ https://aka.ms/vsmarketplace-ToU

Reviewer 2: I haven't read the TOU yet, but you can't link specifically to https://marketplace.visualstudio.com/VSCode because they hide the URL of the file. I tried finding a URL there and found that it's hidden even on the download button.

Reviewer 2: @nickboldt, we can make it explicit by adding a short sentence to say that linking to https://marketplace.visualstudio.com/VSCode cannot and/or should not be attempted. Would that resolve your concern?

Reviewer: I was suggesting we remove the whole tip, since "you can add the extension directly" is not a valid distribution mechanism as it violates the TOU.

Reviewer: also see above,
```
Otherwise the link:https://aka.ms/vsmarketplace-ToU[Terms Of Use] for the link:https://marketplace.visualstudio.com/VSCode[Visual Studio Marketplace] allow using extensions only in Microsoft's products and services.
```

Reviewer 2: OK. @nickboldt, I will remove this TIP then. I hope @svor is okay with this removal.
![image](https://user-images.githubusercontent.com/78564304/208548958-f984f931-593a-4d3e-bf04-415c58e99d48.png)

Author: I don't think we should remove this TIP. Maybe just make it as an another **if** in step 2,  because it could be useful when the extension (vsix file) is published on some custom openvsx registry, or on GitHub or on some web server.

Reviewer: how about: 

```
If the extension author is not cooperative, you can add the extension directly from the 
`.vsix` file, as long as the license and terms of use allow it, and the file can be accessed 
from a https:// URL while building the plugin registry container.
```

?

Author: LGTM

[source,json,subs="+quotes"]
----
    {
        "id": "__<published_by>__.__<unique_identifier>__",
        "download": "__<url_to_download_vsix_file>__",
        "version": "__<extension_version>__"
    }
----
====

** You can remove the extension by deleting its object from the `openvsx-sync.json` file.

. Build the plugin registry container image and publish it to a container registry like link:https://quay.io/[quay.io]:

.. {empty}
+
[subs="+attributes,+quotes"]
----
$ ./build.sh -o __<username>__ -r quay.io -t custom
----

.. {empty}
+
[subs="+attributes,+quotes"]
----
$ {docker-cli} push quay.io/__<username/plugin_registry:custom>__
----

> . Edit the `CheCluster` custom resource in the cluster to point to the image (for example, on link:https://quay.io/[quay.io]) and then save the changes:

Reviewer: Should we note that on subscription update, CheCluster customizations will be lost? 

@svor am I correct in this assumption, or do CheCluster changes like this survive subscription update?

AFAIU the trickledown is Subscription -> Operator Bundle -> CSV -> related images. 

So if the subscription updates from 3.3 -> 3.4, customizations to a 3.3 deployment will be lost after OLM updates the installation to 3.4. Then an admin would have to re-update the CheCluster w/ the new image reference.

Right?

Author: @nickboldt yes, you are right after the migration 3.3 -> 3.4, the image will be updated

Reviewer: added a warning in https://github.com/eclipse-che/che-docs/pull/2503/files#r1053258086 

WDYT @max-cx @svor ?

** If the extension is published on link:https://open-vsx.org[open-vsx.org], you can add it by including an object with the extension `id` in the format __<published_by>__.__<unique_identifier>__ that you can collect from the extension page on link:https://open-vsx.org[open-vsx.org]:
+
[source,json,subs="+quotes"]
----
    {
        "id": "__<published_by>__.__<unique_identifier>__"
    }
----
+
[TIP]
====
The latest extension version on link:https://open-vsx.org[open-vsx.org] is the default. Alternatively, you can add `"version": "__<extension_version>__"` on a new line to specify a version.
====

** If it is an extension from link:https://marketplace.visualstudio.com/VSCode[Visual Studio Marketplace] that is not available on link:https://open-vsx.org[Open VSX] yet, you can ask the extension author to publish it on link:https://open-vsx.org[open-vsx.org] according to the link:https://github.com/eclipse/openvsx/wiki/Publishing-Extensions#how-to-publish-an-extension[instructions] so that you can add it as an extension that is published on link:https://open-vsx.org[open-vsx.org].
+
[TIP]
====

> If the extension author is not cooperative, you can add the extension directly from the `.vsix` file, as long as the license and terms of use allow it, and the file can be accessed from a https:// URL while building the plugin registry container.

Reviewer 2: @nickboldt and @svor, the only two places where I could find `.vsix` files for download were https://marketplace.visualstudio.com/ and https://open-vsx.org/, and neither of them provides clear URLs that I could use for downloading `.vsix` files. The only scenario that comes to mind is the possibility of downloading `.vsix` files from https://marketplace.visualstudio.com/ and uploading or storing them elsewhere for internal purposes, which would probably constitute a violation of Microsoft's TOU, right?

Additionally, it might not be strategically wise in the long run to get customers into the habit of circumventing the https://open-vsx.org/ registry expansion with such easy fixes as this one.

So I wonder if I am missing some other use case that does not violate Microsoft's TOU. If not, then maybe we should indeed remove this TIP, as Nick suggested previously? If no objections, then I propose replacing this TIP with the TIP in Nick's previous comment ^. WDYT?

Reviewer: There are 4 user stories:

* extension is already on Open VSX: **OK to use**

* extension is on MS VSC marketplace: **NEVER OK to use**. See TOU. 

* I (on behalf of my company) built my own extension, but didn't publish it to a public site because I'm into closed source or it's got prioprietary stuff for my company, so it's on an internal company website from which it can be fetched into the registry container build: not great Open Source ethics, but as long as the license is respected, **this is valid**

* I want something from MS VSC marketplace so I reached out to the publisher and they agreed to publish it to Open VSX: as long as the publisher provided it to Open VSX and we use it from THAT URL... **OK to use**. Perhaps we link to https://github.com/marketplace/actions/publish-vs-code-extension to make it super easy? 

>  some other use case that does not violate Microsoft's TOU

Please read the TOU. It's pretty concise in that you can ONLY use the marketplace w/ MS products. Period. End of story.

Reviewer 2: OK, thank you, Nick. Let me have one more try at it.

:_content-type: PROCEDURE

[id="adding-or-removing-vscode-extensions-in-the-embedded-open-vsx-registry-instance"]
= Adding or removing pass:[<!-- vale RedHat.Spelling = NO -->] pass:[<!-- vale RedHat.Spelling = NO -->] VS Code pass:[<!-- vale RedHat.Spelling = YES -->] pass:[<!-- vale RedHat.Spelling = YES -->] extensions in the embedded Open VSX registry instance

> You can add or remove your choice of pass:[<!-- vale RedHat.Spelling = NO -->] pass:[<!-- vale RedHat.Spelling = NO -->] VS Code pass:[<!-- vale RedHat.Spelling = YES -->] pass:[<!-- vale RedHat.Spelling = YES -->] extensions in the embedded Open VSX registry instance that is cached by {prod-short} to support air-gapped, offline, and proxy-restricted environments.

Reviewer 3: ```suggestion
You can add or remove your choice of pass:[<!-- vale RedHat.Spelling = NO -->] pass:[<!-- vale RedHat.Spelling = NO -->] VS Code pass:[<!-- vale RedHat.Spelling = YES -->] pass:[<!-- vale RedHat.Spelling = YES -->] extensions in the embedded Open VSX registry instance that is deployed by {prod-short} to support air-gapped, offline, and proxy-restricted environments.
```

This will create a custom build of the Open VSX registry, which can be used in your organization’s workspaces. 

> NOTE: A subscription update reverts your `CheCluster` to default settings, and you must reapply your changes to `CheCluster`. 

Reviewer: ```suggestion
NOTE: Subscription updates for the {prod-short} operator will revert any customizations done to the `CheCluster` to the published settings for the updated operator. As such, you must **reapply your changes** to the `CheCluster` object after the update is complete.  You may also choose to build a new registry container based on the updated {prod-short} container, in order to get the latest security fixes and updates. 
```

Reviewer 2: @nickboldt, I'd like to tweak your suggestion. Is the following okay?
 
```suggestion
WARNING: Subscription updates for the {prod-short} Operator revert any customizations of the `CheCluster` custom resource to the published settings for the updated operator. After the update is complete, you must reapply your changes to the `CheCluster` custom resource. You may also choose to build a new registry container based on the updated {prod-short} container, in order to get the latest security fixes and updates.
```

> WARNING: Subscription updates for the {prod-short} Operator revert any customizations of the `CheCluster` custom resource to the published settings for the updated operator. After the update is complete, you must reapply your changes to the `CheCluster` custom resource. You may also choose to build a new registry container based on the updated {prod-short} container, in order to get the latest security fixes and updates.

Reviewer 3: "published settings" is not clear? Do you mean CheCluster factory default settings?
And anyway the CheCluster overriding after a subscription update doesn't look specific to the plugin-registry: why is it here then? The last sentence is the only important one in this context.

```suggestion
WARNING: After a {prod-short} update, you may choose to build a new registry container based on the updated {prod-short} container, in order to get the latest security fixes and updates.
```

Reviewer: but you're not warning admins that after a subscription update they'll lose their configuration/customizations and will have to re-patch the CheCluster again. Shouldn't we let people know this is not a sticky change?

Reviewer 3: After discussing with @nickboldt, we should not provide any warning about losing customizations as upgrading Che doesn't revert the CheCluster changes made by an admin.

.Procedure

include::example$snip_{project-context}-steps-to-add-or-remove-extensions-in-openvsx-sync-json.adoc[]

> ** If the extension is published on link:https://open-vsx.org[open-vsx.org], you can add it by including an object with the extension `id` in the format __<published_by>__.__<unique_identifier>__ that you can collect from the extension page on link:https://open-vsx.org[open-vsx.org]:

Reviewer: ```suggestion
** If the extension is published on link:https://open-vsx.org[open-vsx.org], you can add its `id` in the format __<published_by>__.__<unique_identifier>__. You can find the correct `id` from the extension's listing page on link:https://open-vsx.org[open-vsx.org].
```

Reviewer 2: @nickboldt, I'd like to tweak your suggestion. Is this okay? 
```suggestion
** If the extension is published on link:https://open-vsx.org[open-vsx.org], you can add its extension `id` in the format __<published_by>__.__<unique_identifier>__. You can find the `id` on the extension's listing page on link:https://open-vsx.org[open-vsx.org].
```

Reviewer: since we only care about the `id`, why make the statement more vague with `these details` when the only detail you want is the `id` ?

Reviewer: also "add it by including" could me simplified to "include it" or "add it" :D

+
[source,json,subs="+quotes"]
----
    {
        "id": "__<published_by>__.__<unique_identifier>__"
    }
----
+
[TIP]
====
The latest extension version on link:https://open-vsx.org[open-vsx.org] is the default. Alternatively, you can add `"version": "__<extension_version>__"` on a new line to specify a version.
====

> ** If the extension is only available from link:https://marketplace.visualstudio.com/VSCode[Microsoft Visual Studio Marketplace], but not link:https://open-vsx.org[Open VSX], you can ask the extension publisher to also publish it on link:https://open-vsx.org[open-vsx.org] according to the link:https://github.com/eclipse/openvsx/wiki/Publishing-Extensions#how-to-publish-an-extension[instructions] and the link:https://github.com/marketplace/actions/publish-vs-code-extension[GitHub action].

Reviewer: ```suggestion
** If the extension is only available from link:https://marketplace.visualstudio.com/VSCode[Microsoft Visual Studio Marketplace], but not link:https://open-vsx.org[Open VSX], you can ask the extension publisher to also publish it on link:https://open-vsx.org[open-vsx.org] according to these link:https://github.com/eclipse/openvsx/wiki/Publishing-Extensions#how-to-publish-an-extension[instructions], potentially using this link:https://github.com/marketplace/actions/publish-vs-code-extension[GitHub action].
```

+
[TIP]
====

> * If the extension publisher is not available for publishing the extension on link:https://open-vsx.org[open-vsx.org], and if there is no Open VSX equivalent of the extension, consider link:https://github.com/open-vsx/publish-extensions/issues[notifying] the Open VSX team.

Reviewer: ```suggestion
* If the extension publisher is unavailable or unwilling to publish the extension to link:https://open-vsx.org[open-vsx.org], and if there is no Open VSX equivalent of the extension, consider link:https://github.com/open-vsx/publish-extensions/issues[reporting an issue] to the Open VSX team.
```

* If you have a closed-source extension or an extension developed only for internal use in your company, you can add the extension directly from a `.vsix` file that is accessible from a URL while building the plugin registry container:
+
[source,json,subs="+quotes"]
----
    {
        "id": "__<published_by>__.__<unique_identifier>__",
        "download": "__<url_to_download_vsix_file>__",
        "version": "__<extension_version>__"
    }
----
====
+

> WARNING: The link:https://aka.ms/vsmarketplace-ToU[Microsoft Visual Studio Marketplace Terms of Use] do not permit downloading extensions from link:https://marketplace.visualstudio.com/VSCode[Microsoft Visual Studio Marketplace] for use in Microsoft Visual Studio Code - Open Source that is loaded in your organization's workspaces.

Reviewer: ```suggestion
NOTE: Please read the link:https://aka.ms/vsmarketplace-ToU[Terms of Use] for the link:https://marketplace.visualstudio.com/VSCode[Microsoft Visual Studio Marketplace] before using its resources.
```

Reviewer 2: @nickboldt:

1. I phrased the first sentence the way I did for a reason: My sentence says that customers are not permitted to download extensions from [VS Marketplace]. Your suggested sentence reads that extensions from [VS Marketplace]  cannot be used [at all] because of the Terms of Use. But, correct me if I'm wrong, those same extensions can also be published by their respective authors or forkers elsewhere and downloaded from elsewhere. So I tried avoiding a phrasing similar to yours because those extensions, if concurrently published elsewhere, are effectively available to our customers but just cannot be downloaded from [VS Marketplace] in particular. Don't you think?

2. I find the second sentence in your suggestion confusing for a few reasons: (A) _Only Microsoft products are permitted_ is confusing because we say we offer Microsoft Visual Studio Code - Open Source. (B) You mention a URL, but it takes some thinking to figure out which URL you mean. (C) _Only Microsoft products are permitted to use_ sounds to me like those products are the factual users who are permitted to do something, and could it be an anthropomorphism?

And my last point: :D Since it was Microsoft that composed that restrictive license (i.e., Microsoft does not allow using those extensions in Microsoft Visual Studio Code - Open Source), is it now our job to sugarcoat it?

BTW, I copied the TOU title for the link as it is in Microsoft's original PDF doc. So however awkward it may read, it is from the original text.

Can you live with my text as it is, without changes?

![image](https://user-images.githubusercontent.com/78564304/209223575-8da59325-f9c2-4667-b58f-d8c6e0ec8c11.png)

Reviewer: > Your suggested sentence reads that extensions from [VS Marketplace] cannot be used [at all] because of the Terms of Use.

Yes. That is correct.

If a maintainer chooses to publish the extension to another marketplace (eg., open-vsx) then that's a DIFFERENT URL governed by a DIFFERENT TOU. 

TL;DR: DO NOT DOWNLOAD from M$ marketplace UL or face the wrath of their lawyers. :D 

Downloading from another URL could be OK, if that URL has a more lenient TOU.

Reviewer: re: 2

"Microsoft Visual Studio Code - Open Source" is not a MS product.

It's a pile of open source code, which WE build into a product. It is governed by a different license and different TOU.

Reviewer: I still prefer my statement as it more accurately explains that we cannot willfully download from the MS marketplace, for any reason, in a RH or IBM product. (Doesn't have to be MS VS Code - Open Source, could also be theia, IntelliJ IDEA, or any other IDE or product not sanctioned by the TOU.)

```
The Terms of Use for the Microsoft Visual Studio Marketplace prevent using extensions 
from that URL in {prod-short}. Only Microsoft products are permitted to use files from 
the Microsoft Visual Studio Marketplace.
```

The TOU does NOT mention MS VS Code - Open Source in its list of limited allowed products, which to me implies that the Open Source repo != a product. 

> "Microsoft Visual Studio, Visual Studio for Mac, Visual Studio Code, GitHub Codespaces, Azure DevOps, Azure DevOps Server, and successor products and services (the “In-Scope Products and Services”) offered by us and GitHub, Inc. (“GitHub”).

Reviewer 3: I would just suggest Che admins and users to be mindful of the Microsoft Marketplace ToU that limits access to its extensions. I would not spend time interpreting the ToU or even copying it's content in Che documentation as it can change.

Reviewer: ```
Please read the >Terms of Use< for the >Microsoft Visual Studio Marketplace< before 
using its resources.
```

Something simple like that?

Reviewer 3: Yes I like that.

Reviewer: suggestion updated to a `NOTE` instead of a `WARNING`. @max-cx are you OK with this new simpler pointer to the ToU and passing the action to the end user, rather than interpreting the ToU here?

> ** You can remove the extension by deleting its object from the `openvsx-sync.json` file.

Reviewer: ```suggestion
** You can remove the extension by deleting it from the `openvsx-sync.json` file.
```
